packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,285 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+/**
+ * Inspired by:
+ * - https://github.com/pie-dao/vested-token-migration-app
+ * - https://github.com/Uniswap/merkle-distributor
+ * - https://github.com/balancer-labs/erc20-redeemable
+ *
+ * @title MerkleDistributor contract.
+ * @notice Allows an owner to distribute any reward ERC20 to claimants according to Merkle roots. The owner can specify
+ *         multiple Merkle roots distributions, each of which has its own start time, constraining when claims can be
+ *         executed, and end time, controlling the rate at which a claim is vested.
+ */
+
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "../../common/implementation/Lockable.sol";
+import "../../common/implementation/Testable.sol";
+
+contract MerkleDistributor is Ownable, Lockable, Testable {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    // A Window maps a Merkle root to a reward token address and a claim window.
+    // Claims for this window cannot take place before start time.
+    struct Window {
+        // Claims for this window can begin after `start`.
+        uint256 start;
+        // Claims become fully vested after `end`.
+        uint256 end;
+        // Merkle root describing the distribution.
+        bytes32 merkleRoot;
+        // Currency in which reward is processed.
+        IERC20 rewardToken;
+        // Total amount of rewards distributed this window. This is not enforced
+        // but might be useful to query.
+        uint256 totalRewardsDistributed;
+        // Owner can set this to true to block claims for this window.
+        bool locked;
+    }
+
+    // Represents an account's claim for `amount` within the Merkle root located at the `windowIndex`.
+    struct Claim {
+        uint256 windowIndex;
+        uint256 amount;
+        address account;
+        bytes32[] merkleProof;
+    }
+
+    // Windows are mapped to arbitrary indices.
+    mapping(uint256 => Window) public merkleWindows;
+
+    // Keep track of the amount an account has claimed from a window.
+    mapping(uint256 => mapping(address => uint256)) public amountClaimedFromWindow;
```

Comment:
So this enables vesting but not partial claims?

Reply:
Correct but it also could be used to allow partial claims. See [this discussion](https://github.com/UMAprotocol/protocol/pull/2598#discussion_r583924839) where you'd need to pass an additional `partialAmount` param into the `claimWindow()` method, and the changed logic in the claim method would be to set `amountClaimedFromWindow = min(partialAmount, vestedAmount)` instead of `amountClaimedFromWindow = vestedAmount`